Src/Fido2/AttestationFormat/Apple.cs

```diff
@@ -0,0 +1,148 @@
+﻿using System;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using Asn1;
+using Fido2NetLib.Objects;
+using PeterO.Cbor;
+
+namespace Fido2NetLib.AttestationFormat
+{
+
+    internal class Apple : AttestationFormat
+    {
+        private readonly IMetadataService _metadataService;
+        public Apple(CBORObject attStmt, byte[] authenticatorData, byte[] clientDataHash, IMetadataService metadataService)
+            : base(attStmt, authenticatorData, clientDataHash)
+        {
+            _metadataService = metadataService;
+        }
+
+        public override void Verify()
+        {
+            // Reference: https://developer.apple.com/documentation/devicecheck/validating_apps_that_connect_to_your_server
+
+            // Verify that attStmt is valid CBOR conforming to the syntax defined above and
+            // perform CBOR decoding on it to extract the contained fields.
+            if (0 == attStmt.Keys.Count || 0 == attStmt.Values.Count)
+                throw new Fido2VerificationException("Attestation format apple must have attestation statement");
+
+            if (null == Alg || true != Alg.IsNumber)
+                throw new Fido2VerificationException("Invalid apple attestation algorithm");
+
+            // Verify that x5c is present
+            if (null == X5c || CBORType.Array != X5c.Type || 0 == X5c.Count || null != EcdaaKeyId)
+                throw new Fido2VerificationException("Malformed x5c array in apple attestation statement");
+
+            // Validate certificate lifetimes
+            var enumerator = X5c.Values.GetEnumerator();
+            while (enumerator.MoveNext())
+            {
+                if (null == enumerator || null == enumerator.Current
+                    || CBORType.ByteString != enumerator.Current.Type
+                    || 0 == enumerator.Current.GetByteString().Length)
+                    throw new Fido2VerificationException("Malformed x5c cert found in apple attestation statement");
+
+                var x5ccert = new X509Certificate2(enumerator.Current.GetByteString());
+
+                // X509Certificate2.NotBefore/.NotAfter return LOCAL DateTimes, so
+                // it's correct to compare using DateTime.Now.
+                if (DateTime.Now < x5ccert.NotBefore || DateTime.Now > x5ccert.NotAfter)
+                    throw new Fido2VerificationException("Apple signing certificate expired or not yet valid");

```

Comment:
Add the not-before and not-after times to the exception message, for easier debugging?

Reply:
Also, doesn't this loop over all certificates in the chain? If so, technically it may be the CA certificate (or an intermediate certificate) that is invalid.

Reply:
Again, copied from upstream, but it is supposed to check only the first certificate, which contains the attestation public key itself. That will always be checked, while the rest is checked only if we are checking the certificate chain.

We could add the dates in the error message, but this would again be a deviation from the way this is done for other formats.

Reply:
My previous comment was incorrect. Updating error message.